pkg/policycache/cache_test.go

```diff
@@ -1045,3 +1153,34 @@ func Test_NsMutate_Policy(t *testing.T) {
 	}
 
 }
+
+func Test_Validate_Enforce_Policy(t *testing.T) {
+	pCache := newPolicyCache(log.Log, dummyLister{}, dummyNsLister{})
+	policy1 := newValidateAuditPolicy(t)
+	policy2 := newValidateEnforcePolicy(t)
+	pCache.Add(policy1)
+	pCache.Add(policy2)
+
+	validateEnforce := pCache.get(ValidateEnforce, "Pod", "")
+	if len(validateEnforce) != 2 {
+		t.Errorf("adding: expected 2 validate enforce policy, found %v", len(validateEnforce))
+	}
+
+	validateAudit := pCache.get(ValidateAudit, "Pod", "")
+	if len(validateAudit) != 0 {
```

Comment:
For both policies, it's `Enforce` for `default` namespace and `Audit` for `test` namespace, why do we expect 0 audit policy here?

How is `check-label-app-audit` policy applied for `test` namespace?

Reply:
Because we're testing the following here: if the policy has Enforce for any of the namespaces, we treat it as an Enforce Policy and add it to the cache.

> How is `check-label-app-audit` policy applied for `test` namespace?
In Audit mode.

Reply:
Then when the audit handler gets audit policies, there's no such policy returns? Can you attach a test result for this?

Reply:
It does, but it's not fetching them from the Policy Cache. From your comment earlier:
```
	ws.auditHandler.Add(request.DeepCopy())
```

Reply:
We do get the audit policies when the audit handler processes the request, why it's not fetching?
```
	policies := h.pCache.GetPolicies(policycache.ValidateAudit, request.Kind.Kind, request.Namespace)
```

>**Can you attach a test result for this?**

Reply:
Understood! The test `Test_Add_Validate_Audit` does cover it.

Reply:
I see what you mean. Just tested with this policy:
```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: check-label-app-test
spec:
  validationFailureAction: audit
  validationFailureActionOverrides:
    - action: enforce
      namespaces:
        - default
    - action: audit
      namespaces:
        - test
  rules:
    - name: check-label-app
      match:
        resources:
          kinds:
          - Pod
      validate:
        message: "The label `app` is required."
        pattern:
          metadata:
            labels:
              app: "?*"
```
In case `validationFailureAction: audit` and `validationFailureActionOverrides` has `action: enforce` defined, the request is not going to the Audit Handler because we add the Policy to the PolicyCache using the `ValidateEnforce` key. However, since we use the same function in both the Handlers to validate (`handleValidation()`) and there we check `blocked := toBlockResource(engineResponses, logger)` which comes to be false in case of Audit, the validation failure is logged as expected.

Reply:
The fact is that as long as `enforce` behavior is defined in the policy, no matter it's in `validationFailureAction` or `validationFailureActionOverrides`, we'll store it as the ValidateEnforce Policy.

Take the following policy as an example, it's `enforce` across all namespaces except in namespace `test`. And for pods in `test` namespace, the policy is actually applied in the "enforce" validate webhook, not in the audit handler. It should be fine for small clusters but could compromise the performance for large clusters. 
```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: check-label-app-test
spec:
  validationFailureAction: enforce
  validationFailureActionOverrides:
    - action: enforce
      namespaces:
        - default
    - action: audit
      namespaces:
        - test
  rules:
    - name: check-label-app
      match:
        resources:
          kinds:
          - Pod
      validate:
        message: "The label `app` is required."
        pattern:
          metadata:
            labels:
              app: "?*"
```
```
➜  ✗ k run pod --image=nginx             
Error from server: admission webhook "validate.kyverno.svc-fail" denied the request: 

resource Pod/default/pod was blocked due to the following policies

check-label-app-test:
  check-label-app: 'validation error: The label `app` is required. Rule check-label-app
    failed at path /metadata/labels/app/'

➜  ✗ k -n test run pod --image=nginx
pod/pod created

➜  ~ k get polr -n test
NAME           PASS   FAIL   WARN   ERROR   SKIP   AGE
polr-ns-test   0      1      0      0       0      0s
```

Reply:
Agreed. Shall we create a separate enum for the policies that perform both actions?

Reply:
Can you elaborate more? How should we store and apply the policy?

Reply:
I was thinking, we use a separate Policy Type key (say ValidateEnforceAudit) for policies that use Override. Though, this might not be necessary.

We can modify the `GetPolicies()` method to check against the request's Namespace for the Cluster Policies that are using Override.
For instance, considering the above policy as an example, if we have a resource coming in with its namespace set to `default` then we'll filter this resource out if the parameter `pkey=ValidateAudit` (as called by the Audit Handler).
Similarly, if the resource has `namespace: test` and `pkey=ValidateEnforce` then we'll remove this Policy from the `policies` slice.

Reply:
I get it now. We probably need to improve the policy cache for some granular queries. 

Can you please capture your thoughts in a separate issue and add it to 1.6.1?

Reply:
Sure!